
{% extends 'happy_vcfeval/base.html' %}
{% block content %}
{% load widget_tweaks %}



    <div class="container" style="padding-top:5px;text-align:justify">
      {% if known_issues %}
		<div class="panel panel-danger">
		  <div class="panel-heading">
		  	<span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span>
		  	Known issues...
		  </div>
		  <div class="panel-body">
		  	<ul>
		  	{% for line in known_issues %}
		    	<li>{{ line | safe }}</li>
		    {% endfor %}
		    </ul>
			</div>
		</div>
	  {% endif %}
      {% if new_features %}
    <div class="panel panel-success">
      <div class="panel-heading">
        <span class="glyphicon glyphicon-ok-sign" aria-hidden="true"></span>
        Recent updates...
      </div>
      <div class="panel-body">
        <ul>
        {% for line in new_features %}
          <li>{{ line | safe }}</li>
        {% endfor %}
        </ul>
      </div>
    </div>
    {% endif %}
      <!-- Example row of columns -->
      <div class="row">
      	<div class="col-md-6">
        <div class="col-md-12" style="height:965px;background-color:#EEE;border-radius:10px">
          <div style="padding:20px">
          <h2>About</h2>
          <p>This tool is provided for clinical labs to assess the performance of their NGS workflows. It uses Illumina hap.py {{happy_version}}, powered by RTG vcfeval, to compare a sample VCF to the NIST Genome in a Bottle NA12878 truth set. This is based on the precisionFDA workflow and follows GA4GH recommendations for benchmarking. See links below for more information.
          </p> 
          <p>You can use this tool to assess NA12878 sequencing data you have generated by supplying a VCF file containing your variants and a BED file to restrict analysis to regions covered by your panel.
          </p>
          <p>Alternatively if you just want to assess your bioinformatics pipeline, a pair of FASTQ files are provided. These were generated from Illumina NextSeq paired end whole exome sequencing on the  NA12878 sample. You can run these through your pipeline and upload the resulting VCF (no BED file necessary). 
          </p>
          <p>
          You will receive an email containing the output from hap.py, which includes a summary file containing recall (sensitivity) and precision, as well as more detailed results. See the links below for more detail about interperting these results.
          </p>
          <h2>Useful Info</h2>
          <ul>
          <li><a href="https://precision.fda.gov/challenges/truth/results">PrecisionFDA guide to processing and results</a></li>
          <li><a href="https://github.com/ga4gh/benchmarking-tools/tree/master/doc/ref-impl">GA4GH illustration of workflow</a></li>
          <li><a href="https://github.com/Illumina/hap.py">Illumina hap.py</a></li>
          <li><a href="https://github.com/RealTimeGenomics/rtg-tools">RTG Tools vcfeval</a></li>
          </ul>
          <h2>Our Results</h2>
          <p>As a reference point we are provinding our results for the provided FASTQ files, which we ran through our pipeline using BWA MEM and GATK Best Practices. The summary is shown below, and the full result set can be downloaded <a href=" {{our_results}} ">here</a>.</p>
          <table class=table>
          <tr>
          	<th>Type</th>
          	<th>Recall (Sensitivity)</th>
          	<th>Precision</th>
          </tr>
          <tr>
          	<td>SNPs</td>
          	<td>0.994763</td>
          	<td> 0.996936</td>
          </tr>
          <tr>
          	<td>INDELs</td>
          	<td>0.91738</td>
          	<td>0.907096</td>
          </tr>
          </table>

          <h2>Contact</h2>
          <p>If you have any questions or comments, please post in <a href="https://binfx.slack.com/"><b>Slack</b></a></p>
          </div>
        </div>
        </div>
        <div class="col-md-6">
        <div class="col-md-12" style="height:965px;background-color:#EEE;border-radius:10px"">
          <div style="padding:20px">
          <h2>Instructions</h2>
          <ol>
          <li>(OPTIONAL) Download paired end FASTQ files (~12GB): <a href=" {{na12878_fastq}} ">https://stickie.be{{na12878_fastq}}</a></li>
          <li>Process above FASTQs or your own NA12878 data through your pipeline.</li>
          <li>Submit the gzipped VCF file produced by your pipeline, along with your email address and an optional BED file to restrict analysis regions, using the form below. 
            <b>Files should not contain patient identifiers.</b></li>
          <li>Await an email containing a link to your results. <br><b>WARNING:</b> Email might be marked as spam, please whitelist gst-tr.mokaguys@nhs.net!!</li></li>
          </ol>
			<h2>Submit</h2>
			<form method="post" enctype="multipart/form-data">
			{% csrf_token %}
			{% if form.non_field_errors %}
			<div class="alert alert-warning">
			  {{ form.non_field_errors }}
			</div>
			{% endif %}
			<div class="form-group">
			 <label for="email">{% render_field form.email.label %}</label>
			{% render_field form.email class="form-control"%}
			<p class="help-block">Results will be sent to this address</p>
			</div>
			 {% if form.email.errors %}
			<div class="alert alert-warning">
			  {{ form.email.errors }}
			</div>
			{% endif %}
			<div class="form-group">
			 <label for="vcf_file">{% render_field form.vcf_file.label %}</label>
			{% render_field form.vcf_file %}
			<p class="help-block">Max file size: 200MB</p>
			</div>
			 {% if form.vcf_file.errors %}
			<div class="alert alert-warning">
			  {{ form.vcf_file.errors }}
			</div>
			  {% endif %}
			<div class="form-group">
			 <label for="bed_file">{% render_field form.bed_file.label %}</label>
			{% render_field form.bed_file %}
			<p class="help-block">Max file size: 200MB</p>
			</div>
			 {% if form.bed_file.errors %}
			<div class="alert alert-warning">
			  {{ form.bed_file.errors }}
			</div>
			  {% endif %}
			  <button type="submit" class="btn btn-default">Submit</button>
			  <p class="help-block">PLEASE NOTE: Upload may be slow if large VCF files are being submitted. Please be patient!</p>
            </form></p>
            </div>
          </div>
       </div>
      </div>
    </div> <!-- /container -->

{% endblock %}